### PR TITLE
Added DropNullColumn transformer to remove columns that contain only nulls 

### LINKED ISSUE #1110: Add a `DropEmpty` flag to `tabular_learner`
### Problem Description

I am currently working with tables that may contain columns that contain only null/missing values. I also need to perform imputation (with sklearn's SimpleImputer) over all the columns to fill nulls there. 

A simplified version of my pipeline is this:

```py
import polars as pl
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from skrub import TableVectorizer


df_ = pl.DataFrame({"a": [1,2,3], "empty":[None, None, None]})

num_transformer = Pipeline(
    steps=[
        ("imputer", SimpleImputer(strategy="mean", keep_empty_features=False)),
        ("scaler", StandardScaler()),
    ]
)

# Transformer for high cardinality categorical features 
high_card_transformer = Pipeline(
    steps=[
        (
            "imputer",
            SimpleImputer(strategy="most_frequent", keep_empty_features=False),
        ),
        ("onehot_encoder", OneHotEncoder(handle_unknown="ignore", sparse_output=False)),
    ],
)

low_card_transformer = Pipeline(
    steps=[
        (
            "imputer",
            SimpleImputer(strategy="most_frequent", keep_empty_features=False),
        ),
        (
            "onehot_encoder",
            OneHotEncoder(handle_unknown="ignore", sparse_output=False),
        ),
    ],
)
preprocessor = TableVectorizer(
    low_cardinality=low_card_transformer,
    high_cardinality=high_card_transformer,
    numeric=num_transformer,
)

preprocessor.fit_transform(df_)

```

This raises an exception on column `empty` because no values are found. 

### Feature Description

Rather than raising an exception at the end of the pipeline, which makes debugging more complex (the error messages were less than helpful), it would be nice to have an automated check of all columns before any operation is done to screen for empty columns. 

Some possible improvements could be:
- A warning to the user ("Column something something is empty") 
- Dropping the column immediately
- Dropping the column if the fraction of null values is larger than a given threshold

This should also be added to the `tabular_learner` so that if such a situation occurs, it doesn't fail for some weird reason. 

### Alternative Solutions

Setting the SimpleImputer parameter `keep_empty_features=True` seems to fix the problem (or at least, the code does not fail anymore). Null values are replaced by 0. 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by rcap107**
fixes #1110

DropNullColumn (provisional name) takes as input a column, and drops it if all the values are nulls or nans. TableVectorizer was also updated with a `drop_null_columns` flag set to `False` by default; if the flag is set to `True`, the DropNullColumn is added as a processing step for all columns. 

I've also added `drop` and `is_all_null` to `_common.py`, though I am not sure if they should go there. Maybe `is_all_null` can stay in the `DropNullColumn` file. 

The test I wrote passes, but I'm not sure if it's good enough. 

The documentation is still missing.

### CODE REVIEW

File: skrub/tests/test_dropnulls.py
Comment (TheooJ) [inline skrub/tests/test_dropnulls.py:10]: I would call the module `_drop_null.py`

File: skrub/tests/test_dropnulls.py
Comment (TheooJ) [inline skrub/tests/test_dropnulls.py:36]: All the tests in this file are going to test DropNullColumn so I'd avoid this test name

File: skrub/_dataframe/_common.py
Comment (TheooJ) [inline skrub/_dataframe/_common.py:830]: You need a test for this in `test_common.py`

File: skrub/_dataframe/_common.py
Comment (TheooJ) [inline skrub/_dataframe/_common.py:836]:
```suggestion
    return col.isna().all()
```

I think it's already a boolean here

File: skrub/_dataframe/_common.py
Comment (TheooJ) [inline skrub/_dataframe/_common.py:1213]: I don’t know if `drop` is necessary, you could directly use skrub selectors: 
`df = s.select(df, ~s.cols(col))`

File: skrub/_table_vectorizer.py
Comment (TheooJ) [inline skrub/_table_vectorizer.py:195]: Do we want it to be `True` by default ?

File: skrub/tests/test_dropnulls.py
Comment (TheooJ) [inline skrub/tests/test_dropnulls.py:9]:
```suggestion
from skrub import _dataframe as sbd
```

File: skrub/tests/test_dropnulls.py
Comment (TheooJ) [reply skrub/tests/test_dropnulls.py:10]: (this is the same thing, we need to uniformize this in the codebase)

File: skrub/tests/test_dropnulls.py
Comment (TheooJ) [inline skrub/tests/test_dropnulls.py:49]: This test needs to go in the TV test file IMO

**Review by TheooJ — COMMENTED** [review]
Hi @rcap107 ! I made a first pass and have a few comments :

- Personnally I like the name `DropNullColumn`, I think it’s clear what it does !
- I would rename the file `_drop_null.py`
- Make sure you `pre-commit run --all-files` before pushing, it seems to be what’s breaking the CI for you here
- I think `is_all_null` could be placed in the `DropNullColumn` file if it’s only used there for now, but I could also see it being in `_common.py`

File: skrub/tests/test_dropnulls.py
Comment (TheooJ) [reply skrub/tests/test_dropnulls.py:9]: (This is the same thing, we need to uniformize the codebase)

**Comment by rcap107** [comment]
Hi @TheooJ, thanks a lot for the comments! I'll address them and update the PR :+1:

File: skrub/_dataframe/_common.py
Comment (rcap107) [reply skrub/_dataframe/_common.py:836]: `col.isna().all()` of type `numpy.bool`, I'm not sure if that makes a difference downstream, I decided to cast it to bool to avoid issues.

File: skrub/_table_vectorizer.py
Comment (rcap107) [reply skrub/_table_vectorizer.py:195]: That should be discussed with others I think

File: skrub/tests/test_dropnulls.py
Comment (rcap107) [reply skrub/tests/test_dropnulls.py:49]: I can move it :+1:

File: skrub/tests/test_drop_nulls.py
Comment (rcap107) [inline skrub/tests/test_drop_nulls.py:60]: Not sure how to write this check so that it works with either pandas or polars

Comment (TheooJ) [reply skrub/tests/test_drop_nulls.py:60]: You could use `df_module` as a fixture in the test by adding it to the arguments, then comparing series instead of numpy arrays:

```
df_module.assert_column_equal(
    sbd.col(drop_null_table, "value_almost_null"),
    df_module.make_column("value_almost_null", ["almost", None, None]),
)
```

Comment (TheooJ) [reply skrub/tests/test_drop_nulls.py:60]: This also circumvents that depending on the version of pandas, null values are not treated the same

**Comment by jeromedockes** [comment]
the failure in the min-deps environment is not related to this pr; the fix is in #1122

File: skrub/_dataframe/_common.py
Comment (jeromedockes) [reply skrub/_dataframe/_common.py:836]: I think that's a good idea

File: skrub/_dataframe/_common.py
Comment (jeromedockes) [inline skrub/_dataframe/_common.py:848]: when it's not all null we want to return False rather than None

File: skrub/_drop_null.py
Comment (jeromedockes) [inline skrub/_drop_null.py:15]:
```suggestion
```

to follow the scikit-learn requirements, estimators must not do anything in their __init__ except store the parameters. here you don't have any parameters to store so you don't need to define __init__

File: skrub/_drop_null.py
Comment (jeromedockes) [inline skrub/_drop_null.py:30]: we want the schema of the output to be the same after every call to `transform`. at the moment, the check is done during transform so we decide each time transform is called, depending on the input, if we want to drop it or not. instead, we want to check if the column passed to `fit` is all null, if yes, we always drop during transform (even if we later get a column with non-null values) if no, we never drop (even if we later get a column full of nulls).

so it would look like:

```python
def fit_transform(self, column, y=None):
    self.drop_ = sbd.is_all_null(column)
    return self.transform(column)

def transform(self, column):
    check_is_fitted(self, "drop_")
    return [] if self.drop_ else column

# __sklearn_is_fitted__ not needed anymore
```

File: skrub/_table_vectorizer.py
Comment (jeromedockes) [inline skrub/_table_vectorizer.py:546]: - we may want to insert it after CleanNullStrings? so that if the column becomes full of nulls after converting `"N/A"` to `null` it will be dropped. also it's not important but your transformer never raises a RejectColumn exception so `allow_reject` has no effect you don't need it here and can leave the default

File: skrub/_dataframe/_common.py
Comment (jeromedockes) [inline skrub/_dataframe/_common.py:843]: there could be a mix of nan and null which I think would fall through the cracks here.

For simplicity I think we could do:

```python
def is_all_null(col): return all(is_null(col))
```

but if you really want the specialization for polars to get the speedup with null count it could be

```python
def _is_all_null_polars(col):
    if col.dtype == pl.Null:
        return True
    if col.null_count() == col.len():
        return True
    return all(is_null(col))
```

File: skrub/_drop_null.py
Comment (jeromedockes) [inline skrub/_drop_null.py:1]: nitpick: can we have the same name for the class, the module and the test module? eg `DropNullColumns`, `drop_null_columns`, `test_drop_null_columns` (or the singular version if you prefer)

File: skrub/_table_vectorizer.py
Comment (rcap107) [inline skrub/_table_vectorizer.py:549]: This is ugly, but I'm not sure how to write this better.

File: skrub/_table_vectorizer.py
Comment (rcap107) [reply skrub/_table_vectorizer.py]: I added it after CleanNullStrings, but I think I did it in an ugly way, maybe it can be fixed

File: skrub/_dataframe/_common.py
Comment (jeromedockes) [inline skrub/_dataframe/_common.py:843]: not sure I understand the comment

File: skrub/_drop_null_column.py
Comment (jeromedockes) [inline skrub/_drop_null_column.py:41]:
```suggestion
        check_is_fitted(self)
```

File: skrub/tests/test_drop_null_column.py
Comment (jeromedockes) [inline skrub/tests/test_drop_null_column.py:45]: in theory not to rely on polars or pandas specific behavior we would write `sbd.col(drop_null_table, "value_nan")` rather than `drop_null_table["value_nan"]`

File: skrub/tests/test_drop_null_column.py
Comment (jeromedockes) [inline skrub/tests/test_drop_null_column.py:60]: not sure I understand the 3 tests above, the transformer does not seem to be used? we just compare a column from the table to a new one we construct

File: skrub/tests/test_drop_null_column.py
Comment (jeromedockes) [inline skrub/tests/test_drop_null_column.py:63]: I understand that you put it here to use the fixture defined in this module but I think this test should probably go in `test_table_vectorizer.py`

File: skrub/_table_vectorizer.py
Comment (jeromedockes) [reply skrub/_table_vectorizer.py:195]: I vote for true by default -- there's nothing we can learn from a completely empty column.

if it is False by default, I think it should be set to True in the `tabular_learner`

File: skrub/_dataframe/_common.py
Comment (rcap107) [reply skrub/_dataframe/_common.py]: it was a leftover from the previous version, I rewrote the comments

File: skrub/tests/test_table_vectorizer.py
Comment (rcap107) [inline skrub/tests/test_table_vectorizer.py:532]: I set this to false to keep the original behavior with no DropNullColumns. Given that the default value is True, should I change the test so that the "default behavior" is what is tested here?

Comment (jeromedockes) [reply skrub/tests/test_table_vectorizer.py:532]: I think it's ok the way you did it

File: skrub/_table_vectorizer.py
Comment (jeromedockes) [inline skrub/_table_vectorizer.py:397]: I wonder if we should have an 'if' in there eg call it 'DropColumnIfNull' to make it clearer that this may be a no-op when reading the list of transformations

Comment (rcap107) [reply skrub/_table_vectorizer.py]: I like the idea, it's definitely clearer. I renamed the object.

File: skrub/_dataframe/tests/test_common.py
Comment (jeromedockes) [inline skrub/_dataframe/tests/test_common.py:565]:
```suggestion
    assert ns.is_all_null(df_module.make_column("all_nan", [np.nan, np.nan, np.nan]))
    assert ns.is_all_null(df_module.make_column("all_nan_or_null", [np.nan, np.nan, None]))
```

File: skrub/_drop_column_if_null.py
Comment (jeromedockes) [inline skrub/_drop_column_if_null.py:17]: can you check the formatting of the other estimators in the library, here the section header would look like 
```
Parameters
----------
```

not super important here as it is a private class, but if we decide to make it public it will be important because sphinx relies on this formatting to produce the reference html documentation

**Review by jeromedockes — APPROVED** [review]
thanks a lot @rcap107 !! LGTM once the last couple of nitpicks are addressed :tada:

**Comment by Vincent-Maladiere** [comment]
Hey @rcap107, you haven't addressed my comment in https://github.com/skrub-data/skrub/issues/1110#issuecomment-2408869551. I'm -1 to drop all nan columns in the table vectorizer by default, as these might be useful information after concatenation.

**Comment by rcap107** [comment]
> Hey @rcap107, you haven't addressed my comment in [#1110 (comment)](https://github.com/skrub-data/skrub/issues/1110#issuecomment-2408869551). I'm -1 to drop all nan columns in the table vectorizer by default, as these might be useful information after concatenation.

I'm fine with having the drop flag false by default, I don't see how having a concatenation of empty columns brings in any information after concatenation :thinking: If the point is acting as indicator, then the user should just add the column as constant, which would be more informative than a null value anyway (and it wouldn't trigger the drop). 

I can imagine a situation where both tables A and B have a column `col_1`, and `col_1` is all null in A, but isn't in B. With the current behavior, `col_1` is dropped from A and kept in B, and if I try to concatenate A and B I'd expect the behavior to be "add a column called `col_1` to the final table and fill the rows of A with nulls", unless an exception is raised if the same set of columns must be provided (which may be the case and I'm just not aware of it). 

In any case I can modify the code to have the three choices rather than a binary flag.

**Comment by Vincent-Maladiere** [comment]
>  then the user should just add the column as constant

My intuition is that we should avoid assuming users would "just do" things and try to prevent this kind of mistake.

> I can imagine a situation where both tables A and B have a column col_1, and col_1 is all null in A, but isn't in B ...

Consider a situation where you have a TableVectorizer in a scikit-learn pipeline, trained in a production environment every week. If you built this pipeline assuming you would have access to some column "col_1", but for some reason, the data production system now only produces NaN values for this column (e.g. this column has been deprecated, which happened quite often), then this pipeline would raise a "column is missing" error that would be tricky to understand, because most users wouldn't be familiar with the internals of the TableVectorizer. Having a clear error or a warning would significantly help debugging and maintenance in such a scenario. WDYT?

**Comment by rcap107** [comment]
Yes, I agree that having a more informative error would be better. 

I am changing the code so that it is possible to set different either `drop`, `ignore` or `raise`. I modified the init of DropColumnIfNull. Sidenote, where should the error checking for that value be? In DropNull or in TableVectorizer, or both? 

I could also add a `warn` strategy, but wouldn't it be better to set it so that it triggers when in debug mode? I am not sure how to do that. 

For the default strategy, we can use `ignore` to keep null columns.

**Comment by Vincent-Maladiere** [comment]
> Sidenote, where should the error checking for that value be? In DropNull or in TableVectorizer, or both?

I think the most practical would be raising a specific error in `DropNull`. The way I picture it:
- We add a new argument in `TableVectorizer`, like `all_nulls_cols` whose options are "warn" (default), "ignore", "drop", "raise"
- `DropNull` then takes the same argument and default

Also, "warn" looks like a good default, but that's up for discussion.

WDYT?

> [...] when in debug mode?

I'm not sure to understand what debug mode is?

**Comment by jeromedockes** [comment]
I think "ignore" is a bit ambiguous -- if I ignore a column does it mean I don't use it / leave it out of the output? so I would suggest "keep" instead

**Comment by rcap107** [comment]
> > [...] when in debug mode?
> 
> I'm not sure to understand what debug mode is?
nevermind


> * `DropNull` then takes the same argument and default
> 
> Also, "warn" looks like a good default, but that's up for discussion.

We also need to decide whether it's "warn and drop", or "warn and keep", and be clear about the behavior.

File: skrub/tests/test_table_vectorizer.py
Comment (rcap107) [inline skrub/tests/test_table_vectorizer.py:779]: This test is still failing because the TableVectorizer is not raising the correct exception and I don't know how to make it do that.

**Comment by rcap107** [comment]
I have updated the code to have "warn and keep" as the default behavior, I think it's the version that makes the most sense. 

At the moment the only problem I have is that I don't know how to raise the proper exception from TableVectorizer. In  DropNull I am raising RejectColumn, but then I don't know how to propagate it correctly,

File: skrub/tests/test_table_vectorizer.py
Comment (jeromedockes) [reply skrub/tests/test_table_vectorizer.py:779]: [here](https://github.com/skrub-data/skrub/pull/1115/files#diff-915f4c6f453724cbd65c06d5c7f55a1aed027f2b5649d15ec4ec3b9edb90ad70R77) raise a ValueError instead of RejectColumn

Comment (jeromedockes) [reply skrub/tests/test_table_vectorizer.py:779]: rejectcolumn is a way to signify to the tablevectorizer "I'm not the right transformer for this column, don't apply me here".

Comment (rcap107) [reply skrub/tests/test_table_vectorizer.py:779]: Gotcha, fixed

**Comment by jeromedockes** [comment]
I find it a bit weird to have a DropColumnIfNull that does not drop the column and just raises an exception. maybe it should be named something like CheckNulls or something?

**Comment by jeromedockes** [comment]
> Consider a situation where you have a TableVectorizer in a scikit-learn pipeline, trained in a production environment every week. If you built this pipeline assuming you would have access to some column "col_1", but for some reason, the data production system now only produces NaN values for this column

I'm not sure I understand -- in any case the tablevectorizer chooses the estimators during fit so the schema of the output can change every week in this scenario. eg if a column has one more unique value than the previous week it can change from a one-hot encoding to a gap encoding. or for example if you had `high_cardinality="drop"` those columns would also be dropped or kept depending on their content

**Comment by jeromedockes** [comment]
so if you want consistent output schema (number of columns, names and types) across training the tablevetorizer is not what you want anyway

**Comment by Vincent-Maladiere** [comment]
> I find it a bit weird to have a DropColumnIfNull that does not drop the column and just raises an exception. maybe it should be named something like CheckNulls or something?

Yes, I agree this sounds weird. I recognize dropping NaN columns –which are usually useless– looks good. What about having the transformer drop by default, but allowing the user to pass other options as arguments in the `TableVectorizer`?

> I'm not sure I understand --  in any case the tablevectorizer chooses the estimators during fit [...]

Right, I was pointing out that even outside of the `TableVectorizer` (e.g. in the `ColumnTransformer`), the `DropColumnIfNull` could raise surprising mistakes.

It's true though that this issue broadly applies to the `TableVectorizer`, hence we should encourage people to perform checks in production (with dedicated check functions? There might be some potential here for production usage).

**Comment by rcap107** [comment]
Following today's discussion with @Vincent-Maladiere and @jeromedockes, I reverted the changes back to the old version (with the simple flag)

**Review by jeromedockes — APPROVED** [review]
LGTM! thanks again @rcap107 

(the codecov report is bogus)

I'll let @Vincent-Maladiere do a final review & merge

**Review by Vincent-Maladiere — COMMENTED** [review]
In our IRL discussion, haven't we agreed on a threshold for the null ratio in the column (which is 1.0 by default)?

Apart from that, and to keep in mind for later, the items we discussed were:
1. The ability to freeze TableVectorizer column-to-transformer mapping. It would help to obtain consistent results for retraining in an automated environment, and having more sensible errors to debug.
2.  Decoupling the check/cleaning part of the TableVectorizer (which comes before the vectorizing part) so that it can be used as a standalone object wherever.

**Comment by jeromedockes** [comment]
> In our IRL discussion, haven't we agreed on a threshold for the null ratio in the column (which is 1.0 by default)?

we definitely did; I had understood it would be tackled in a separate PR

+1 for points 1. and 2. -- let's open a separate issue for 1. and there is #925 for 2.

**Comment by rcap107** [comment]
I also understood that the threshold would be added in a separate issue

**Review by Vincent-Maladiere — APPROVED** [review]
LGTM then, thanks @rcap107 :)

**Comment by jeromedockes** [comment]
yay :tada: !! thanks @rcap107 !

**Comment by rcap107** [comment]
:tada:

**Pull request merged**
